### ISSUE TITLE
Failed to lookup for nodes in docker swarm

### ISSUE THREAD

**Issue by ildarg** [minio/minio#4606]
The minio servers can not be restarted in Docker Swarm. There is "Failed to lookup.." in container's log.

## Expected Behavior

Minio servers should be restarted properly.

## Current Behavior

..Failed to lookup minio2
.Failed to lookup minio3
.Failed to lookup minio4
Failed to resolve hosts.. retrying after 1 second.

## Possible Solution

Starting from the scratch

## Steps to Reproduce (for bugs)

Installed according to https://docs.minio.io/docs/deploy-minio-on-docker-swarm
1.  docker swarm init --advertise-addr <MANAGER-IP> on server 1
2.  docker swarm join --token ..... <MANAGER-IP>:2377 on server 2,3,4
3. on server 1
echo "AKIAIOSFODNN7EXAMPLE" | docker secret create access_key -
echo "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY" | docker secret create secret_key -
4. on server 1
docker stack deploy --compose-file=docker-compose-secrets.yaml minio_stack
5. service docker restart

## Your Environment

os: Ubuntu 16.04.1 LTS
docker: 17.05.0-ce
minio image: minio/minio:RELEASE.2017-06-13T19-01-01Z

**Comment by harshavardhana** [comment]
did you wait for all the servers to start properly before they are being restarted? @ildarg

**Comment by ildarg** [comment]
sure, I waited more then 30 minutes (several times)
PD I did also
1) docker stack rm minio_stack
2) docker stack deploy --compose-file=docker-compose-secrets.yaml minio_stack
no changes

**Comment by nitisht** [comment]
@ildarg can you try removing all the volumes in your stack - `docker volume rm volume_name` (also mentioned in step 4 here: https://docs.minio.io/docs/deploy-minio-on-docker-swarm)

**Comment by harshavardhana** [comment]
> PD I did also
> 
> docker stack rm minio_stack
> docker stack deploy --compose-file=docker-compose-secrets.yaml minio_stack
> no changes

Unfortunately `docker stack rm minio_stack` doesn't delete the volumes it created. You need to remove the volumes.. `docker volume rm volume_name`

**Comment by ildarg** [comment]
I did it after stack removing as desccribed in guide.
PS I hope it is not solution for production mode. I can't delete storage.

**Comment by nitisht** [comment]
> PS I hope it is not solution for production mode. I can't delete storage.

Yes, this is for demo. The problem is Swarm deploys nodes randomly and volumes are not mapped to specific nodes. This is handled by Kubernetes in form of persistent volumes.

**Comment by ildarg** [comment]
Could you please show the way how to fix it ? Do you recommend to work without Docker ?

**Comment by nitisht** [comment]
Hi @ildarg sorry for dropping the ball on this. I looked into Docker Swarm documentation and looks like the way to handle this is to 

- Create bind mounts on Swarm nodes. This way there is a specific path on each host where Minio can store data. Ref: https://docs.docker.com/engine/swarm/services/#give-a-service-access-to-volumes-or-bind-mounts. 

- Then add constraints on specific services to run on specific hosts. For example service `minio1` runs on host `worker1`, service `minio2` runs on host `worker2` and so on. Ref: https://docs.docker.com/compose/compose-file/#placement

Unfortunately Swarm doesn't offer an easy way to handle such scenarios yet. If it is possible, you can also try out Kubernetes based orchestration that handles such scenarios better.

**Comment by nitisht** [comment]
@ildarg please feel free to re-open this if you face the problem again. I am closing this for now.

**Comment by Blaiserman** [comment]
Same thing here with your compose file: https://raw.githubusercontent.com/minio/minio/master/docs/orchestration/docker-swarm/docker-compose-secrets.yaml

I tried to change the volumes to random but died.
```
    volumes:
      - /export
```

**Comment by nitisht** [comment]
@Blaiserman can you please paste the output you're getting and the Minio server version?

**Comment by Blaiserman** [comment]
```
minio/minio:RELEASE.2017-08-05T00-00-53Z
```

```
.Failed to lookup minio2
.Failed to lookup minio3
.Failed to lookup minio4
Failed to resolve hosts.. retrying after 1 second.
.Failed to lookup minio1
.Failed to lookup minio2
.Failed to lookup minio3
.Failed to lookup minio4
Failed to resolve hosts.. retrying after 1 second.
.Failed to lookup minio1
.Failed to lookup minio2
.Failed to lookup minio3
.Failed to lookup minio4
Failed to resolve hosts.. retrying after 1 second.
.Failed to lookup minio1
.Failed to lookup minio2
.Failed to lookup minio3
.Failed to lookup minio4
Failed to resolve hosts.. retrying after 1 second.
.Failed to lookup minio1
.Failed to lookup minio2
.Failed to lookup minio3
```

**Comment by nitisht** [comment]
Is there a way to access your setup @Blaiserman 

Please join our [Slack channel](https://slack.minio.io) for quick discussion.

**Comment by Blaiserman** [comment]
The environment running on my windows laptop on docker toolbox so you can not access to it, but I can answer to your questions.

```
NAME=Boot2Docker
VERSION=17.06.2-ce
ID=boot2docker
ID_LIKE=tcl
VERSION_ID=17.06.2-ce
PRETTY_NAME="Boot2Docker 17.06.2-ce (TCL 7.2); HEAD : ff16afa - Wed Sep  6 00:17:25 UTC 2017"
ANSI_COLOR="1;34"
HOME_URL="http://boot2docker.io"
SUPPORT_URL="https://github.com/boot2docker/boot2docker"
BUG_REPORT_URL="https://github.com/boot2docker/boot2docker/issues"
```

I joined to Slack

**Comment by Blaiserman** [comment]
I created the environment in docker toolbox like this:
```
docker-machine create --driver virtualbox node1
docker-machine create --driver virtualbox node2
docker-machine create --driver virtualbox node3
```

I created a folder for mount on all nodes:
```
docker-machine ssh node1 "sudo rm -rf /mnt/sda1/sda1/minio1"
docker-machine ssh node1 "sudo mkdir /mnt/sda1/minio1"
docker-machine ssh node1 "sudo chown -R docker:docker /mnt/sda1/minio1"

docker-machine ssh node2 "sudo rm -rf /mnt/sda1/minio2"
docker-machine ssh node2 "sudo mkdir /mnt/sda1/minio2"
docker-machine ssh node2 "sudo chown -R docker:docker /mnt/sda1/minio2"

docker-machine ssh node3 "sudo rm -rf /mnt/sda1/minio3"
docker-machine ssh node3 "sudo mkdir /mnt/sda1/minio3"
docker-machine ssh node3 "sudo chown -R docker:docker /mnt/sda1/minio3"

docker-machine ssh node3 "sudo rm -rf /mnt/sda1/minio4"
docker-machine ssh node3 "sudo mkdir /mnt/sda1/minio4"
docker-machine ssh node3 "sudo chown -R docker:docker /mnt/sda1/minio4"
```

Added a tag for the nodes and create the secrets:
```
docker-machine ssh node1 "docker node update node1 --label-add node=node1"
docker-machine ssh node1 "docker node update node2 --label-add node=node2"
docker-machine ssh node1 "docker node update node3 --label-add node=node3"

docker-machine ssh node1 "echo "minio" | docker secret create access_key -"
docker-machine ssh node1 "echo "minio123" | docker secret create secret_key -"
``` 

Lastly this is my working compose file:
```
version: '3.3'

services:
  minio1:
    image: minio/minio:edge
    volumes:
      - /mnt/sda1/minio1:/export
    ports:
      - "9001:9000"
    networks:
      - minio_distributed
    deploy:
      restart_policy:
        delay: 10s
        max_attempts: 10
        window: 60s
      placement:
        constraints:
          - "node.labels.node == node1"
    command: server http://minio1/export http://minio2/export http://minio3/export http://minio4/export
    secrets:
      - secret_key
      - access_key

  minio2:
    image: minio/minio:edge
    volumes:
      - /mnt/sda1/minio2:/export
    ports:
      - "9002:9000"
    networks:
      - minio_distributed
    deploy:
      restart_policy:
        delay: 10s
        max_attempts: 10
        window: 60s
      placement:
        constraints:
          - "node.labels.node == node2"
    command: server http://minio1/export http://minio2/export http://minio3/export http://minio4/export
    secrets:
      - secret_key
      - access_key

  minio3:
    image: minio/minio:edge
    volumes:
      - /mnt/sda1/minio3:/export
    ports:
      - "9003:9000"
    networks:
      - minio_distributed
    deploy:
      restart_policy:
        delay: 10s
        max_attempts: 10
        window: 60s
      placement:
        constraints:
          - "node.labels.node == node3"
    command: server http://minio1/export http://minio2/export http://minio3/export http://minio4/export
    secrets:
      - secret_key
      - access_key

  minio4:
    image: minio/minio:edge
    volumes:
      - /mnt/sda1/minio4:/export
    ports:
      - "9004:9000"
    networks:
      - minio_distributed
    deploy:
      restart_policy:
        delay: 10s
        max_attempts: 10
        window: 60s
      placement:
        constraints:
          - "node.labels.node == node3"
    command: server http://minio1/export http://minio2/export http://minio3/export http://minio4/export
    secrets:
      - secret_key
      - access_key

networks:
  minio_distributed:
    driver: overlay

secrets:
  secret_key:
    external: true
  access_key:
    external: true
```

@workd

**Comment by painslie** [comment]
Hi guys, I started playing with distributed minio on Docker Swarm on September 16th.  I've deployed in several different environments (Debian & Ubuntu) and service providers and I run into this error.

```
.Failed to lookup minio1
.Failed to lookup minio2
.Failed to lookup minio3
.Failed to lookup minio4
Failed to resolve hosts.. retrying after 1 second.
```

That problem was resolved for me by switching from `image: minio/minio` to `image: minio/minio:edge`!  Can't recall who, but someone checked in a change recently for this very issue.
